File: GFG/Python/2025-07-31_Powerful_Integer.py
# GFG POTD - Powerful Integer (Optimal Approach)
# Date: 2025-07-31

from typing import List

class Solution:
    def powerfulInteger(self, intervals: List[List[int]], k: int) -> int:
        # Step 1: Difference map to mark range starts and ends
        mp = {}
        for start, end in intervals:
            mp[start] = mp.get(start, 0) + 1
            mp[end + 1] = mp.get(end + 1, 0) - 1

        # Step 2: Apply sweep line prefix logic
        temp = 0
        result = -1
        for point in sorted(mp.keys()):
            # Before changing temp, check if we're ending a high-frequency range
            if mp[point] < 0 and temp >= k:
                result = point - 1

            temp += mp[point]

            # After applying change, check if we're entering a high-frequency range
            if temp >= k:
                result = point

        return result

File: GFG/C++/2025-07-31_Powerful_Integer.cpp
// GFG POTD - Powerful Integer
// Date: 2025-07-31


// ************************ Brute Force ************************
class BruteSolution {
  public:
    int powerfulInteger(vector<vector<int>>& intervals, int k) {
        int n = intervals.size();
        int maxi = INT_MIN;

        for(int i = 0; i < n; i++) {
            maxi = max(maxi, intervals[i][1]);
        }

        vector<int> arr(maxi + 1, 0);

        for(int i = 0; i < n; i++) {
            int s = intervals[i][0];
            int e = intervals[i][1];
            for(int j = s; j <= e; j++) {
                arr[j]++;
            }
        }

        vector<int> poss;
        for(int i = 0; i <= maxi; i++) {
            if(arr[i] >= k) {
                poss.push_back(i);
            }
        }

        return poss.empty() ? -1 : poss.back();
    }
};

// ************************ Optimal Approach ************************
class OptimalSolution {
  public:
    int powerfulInteger(vector<vector<int>>& intervals, int k) {
        map<int, int> mp;

        for (auto &in : intervals) {
            mp[in[0]]++;
            mp[in[1] + 1]--;
        }

        int temp = 0;
        int res = -1;

        for (auto &it : mp) {
            if (temp >= k && it.second < 0)
                res = it.first - 1;

            temp += it.second;

            if (temp >= k)
                res = it.first;
        }

        return res;
    }
};
